### ISSUE TITLE
The `__module__` attribute is not properly set

### ISSUE THREAD

**Issue by leofang** [cupy/cupy#3266]
```python
>>> import cupy as cp
>>> a = cp.empty(10)
>>> type(a).__module__  # should simply be 'cupy'
'cupy.core.core'
>>> cp.sum.__module__   # should simply be 'cupy'
'cupy.math.sumprod'
>>>
>>> import numpy as np
>>> b = np.empty(10)
>>> type(b).__module__
'numpy'
>>> np.sum.__module__
'numpy'
```
This is needed to be compliant with NumPy, which changed the behavior since 1.16:
> The `__module__` attribute on most NumPy functions has been updated to refer to the preferred public module from which to access a function, rather than the module in which the function happens to be defined. This produces more informative displays for functions in tools such as IPython, e.g., instead of `<function 'numpy.core.fromnumeric.sum'>` you now see `<function 'numpy.sum'>`.

Ref: https://numpy.org/devdocs/release/1.16.0-notes.html#module-attribute-now-points-to-public-modules

**Comment by leofang** [comment]
It seems NumPy uses a `set_module` decorator to overwrite this attribute throughout the codebase:
https://github.com/numpy/numpy/blob/544094aed5fdca536b300d0820fe41f22729ec66/numpy/core/overrides.py#L94-L109

**Comment by leofang** [comment]
Forgot to add: This feature is also useful for other libraries to do NEP-18 (`__array_function__`) dispatch.

**Comment by toslunar** [comment]
Definitely.  Users should see `cupy.ndarray` is printed instead of `cupy.core.core.ndarray`.

For developers, the change will not impact much because there's no `from ... import *` in CuPy.

**Comment by leofang** [comment]
@takagi Question: Looks like this issue is half fixed? The `__module__` attribute on arrays are now shown as "cupy", but the CuPy functions are not yet. Is that right?
